### ISSUE TITLE
How to fix this memory warning?

### ISSUE THREAD

**Issue by peace4j** [emqx/emqx#4049]
There is an alarm "high_system_memory_usage:System memory usage is higher than 70%". 
This alarm happens in a new installed server, low loading or even zero loading. But don't know why it will appear.
In dashboard->Monitor -> Nodes -> Memory info column, it shows "112.48M / 160.51M ".  Is the alarm relevant to this? How to adjust it? No answer found in EMQX online docs.

Best regards,

peace4j

**Comment by tigercl** [comment]
@peace4j Sorry for not replying to you in time. What operating system and emqx version are you using?

**Comment by lddhappy** [comment]
Same problem.

My Enviroment:
Kubernetes Deployment:request 500MB memory and limit 1024MB memory.With low loading or even zero loading.

![image](https://user-images.githubusercontent.com/11736713/105785658-651cb980-5fb6-11eb-89fd-46a205d58756.png)


![image](https://user-images.githubusercontent.com/11736713/105785692-7960b680-5fb6-11eb-912d-e7e60ba7d8ce.png)

**Comment by HJianBo** [comment]
There are three memory values here.
1.  the total memory of the operating system
2. memory requested by Erlang VM
3. memory used by Erlang VM

Currently, this alert is based on the Used/Requested of Erlang VM. It is not based on the ratio of memory already requested by ErlangVM to the total memory of the OS.

**Comment by peace4j** [comment]
Is it a erlang vm issue? 

> 
> 
> @peace4j Sorry for not replying to you in time. What operating system and emqx version are you using?

It's a container running from the official docker emqx images version 4.2.6.
The host box is a 4CPUs/8GBMem vps, running CentOS 7.6 with a kernal 3.10.0-1127.19.1.el7.x86_64. 
Currently there is only one running container on the host.

And don't know when, the alarm disappeared.  No configuration item changed during this period. So, I don't know when it will come back again.

**Comment by peace4j** [comment]
> 
> 
> There are three memory values here.
> 
>     1. the total memory of the operating system
> 
>     2. memory requested by Erlang VM
> 
>     3. memory used by Erlang VM
> 
> 
> Currently, this alert is based on the Used/Requested of Erlang VM. It is not based on the ratio of memory already requested by ErlangVM to the total memory of the OS.

Thanks.
It seems no configuration item in emqx.conf matches ERLang VM memory settings. 
Is it a problem? Do I need to fix it?

**Comment by HJianBo** [comment]
No. There is no configuration to change here. it should be a design problem for us.

**Comment by peace4j** [comment]
> 
> 
> No. There is no configuration to change here. it should be a design problem for us.

Fortunately, it happened occasionally. Now, the connected devices are more than before, but warning disappears. 
Any further progress or information about this issue, please update and let me know. Thanks for your patient reply.

**Comment by ingoratsdorf** [comment]
Apologies for reopening this, but this is still an issue in version 5.5.1 (docker)
I see VM memory use going constantly up and down between 500M and 8G, the rise to 8G is coincidental with high CPU use.
It seems to be happening every 20 sec approx.
See screenshot attached.

Some stats that may help: 10 connections, 234 topics, 246 subscriptions, 8-10 messages/sec.
![Screenshot 2024-04-04 200459](https://github.com/emqx/emqx/assets/3344778/bca8b793-1023-4380-99ee-e33663ac1923)

**Comment by id** [comment]
@ingoratsdorf some quick questions

- do you see anything interesting in the logs?
- do you have any data integrations or web hooks?
- do you have some processing in rule engine?
- is there any pattern to messages, like large or just some special kind of payload every ~20sec?

**Comment by ingoratsdorf** [comment]
Run a trace in EQMX for HA, nothing noteworthy from what I can see.
Along the lines of

`2024-04-04T08:34:36.120390+00:00 [PUBLISH] homeassistant@192.168.1.1:38057 msg: publish_to, topic: cmnd/sunsetlight/led_enableAll, payload: 1
2024-04-04T08:34:36.120566+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBACK(Q0, R0, D0, PacketId=18, ReasonCode=0)
2024-04-04T08:34:36.367324+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=sunsetlight/led_basecolor_rgb/get, PacketId=1877, Payload=FF4300)
2024-04-04T08:34:36.368085+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=sunsetlight/led_dimmer/get, PacketId=1878, Payload=38)
2024-04-04T08:34:36.368328+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=sunsetlight/led_enableAll/get, PacketId=1879, Payload=1)
2024-04-04T08:34:36.371307+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1877, ReasonCode=0)
2024-04-04T08:34:36.410880+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1878, ReasonCode=0)
2024-04-04T08:34:36.411011+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1879, ReasonCode=0)
2024-04-04T08:34:36.617060+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=dishwasher/voltage/get, PacketId=1880, Payload=232.6)
2024-04-04T08:34:36.618130+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1880, ReasonCode=0)
2024-04-04T08:34:37.220099+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q0, R0, D0, Topic=zigbee2mqtt/Ingo's Office Desk, PacketId=undefined, Payload={"child_lock":"UNLOCK","current":0.35,"device":{"applicationVersion":192,"dateCode":"","friendlyName":"Ingo's Office Desk","hardwareVersion":1,"ieeeAddr":"0xa4c1386baf25526e","manufacturerID":4417,"manufacturerName":"_TZ3000_typdpbpg","model":"TS011F_plug_1","networkAddress":25238,"powerSource":"Mains (single phase)","stackVersion":0,"type":"Router","zclVersion":3},"energy":467.04,"indicator_mode":"off/on","last_seen":"2024-04-04T08:34:36.215Z","linkquality":192,"power":116,"power_outage_memory":"restore","state":"ON","update":{"installed_version":192,"latest_version":192,"state":"idle"},"update_available":null,"voltage":231})
2024-04-04T08:34:38.535677+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q0, R0, D0, Topic=zigbee2mqtt/Fridge, PacketId=undefined, Payload={"child_lock":"UNLOCK","current":0.95,"device":{"applicationVersion":192,"dateCode":"","friendlyName":"Fridge","hardwareVersion":1,"ieeeAddr":"0xa4c138a99883c113","manufacturerID":4417,"manufacturerName":"_TZ3000_typdpbpg","model":"TS011F_plug_1","networkAddress":18631,"powerSource":"Mains (single phase)","stackVersion":0,"type":"Router","zclVersion":3},"energy":433.42,"indicator_mode":"off","last_seen":"2024-04-04T08:34:37.527Z","linkquality":124,"power":98,"power_outage_memory":"on","state":"ON","update":{"installed_version":192,"latest_version":192,"state":"idle"},"update_available":null,"voltage":230})
2024-04-04T08:34:38.781518+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=StringLights/voltage/get, PacketId=1881, Payload=218.9)
2024-04-04T08:34:38.781814+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=StringLights/power_apparent/get, PacketId=1882, Payload=56.83)
2024-04-04T08:34:38.784574+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1881, ReasonCode=0)
2024-04-04T08:34:38.784696+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1882, ReasonCode=0)
2024-04-04T08:34:38.883896+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_sent, packet: PUBLISH(Q1, R0, D0, Topic=LivingroomDoor/voltage/get, PacketId=1883, Payload=224.5)
2024-04-04T08:34:38.885795+00:00 [MQTT] homeassistant@192.168.1.1:38057 msg: mqtt_packet_received, packet: PUBACK(Q0, R0, D0, PacketId=1883, ReasonCode=0)`

**Comment by zmstone** [comment]
maybe run below commands (EMQX v5 or newer):
- `emqx ctl retainer info`: this will provide the number of retained messages, when there are a large number of retained messages, subscribing to a wildcard topic without a proper index will cause a table scan which is quite memory intense.

`emqx ctl observer status`: this will render a text buffer like linux `top` command, displaying the top reduction (CPU) processes by default. If you press `m<enter>`, it will  show max RAM usage at the top, if you press `mm<enter>` it will show max RAM increment at the top. you can share a screenshot with us.

**Comment by ingoratsdorf** [comment]
Sorry folks, had to stop the use of EMQX due to those issues and the fact that even when not a single client was connected to it, it was using 20% server CPU. For my purposes, cedalo/mosquitto has a way smaller CPU and memory footprint.
Netdata screenshot attached, the launch of EMQX is abvious. The base system had cedalo/mosquitto already running.
![Screenshot 2024-04-13 135726](https://github.com/emqx/emqx/assets/3344778/35b8a863-253f-43dd-81f8-4034973fecdf)

For completeness, screenshot of observer:
![image](https://github.com/emqx/emqx/assets/3344778/05a85258-c6b6-4aba-b92c-ec68bf032b89)

emqx@072eae475776:/opt/emqx$ emqx ctl retainer info
Number of retained messages: 543

**Comment by zmstone** [comment]
sad to see people leave.
the "smp:4:1" is alarming, it means the server started with 4 schedulers but only has 1 online. 
it's because the poor vm only got one CPU core.

**Comment by kleini** [comment]
I have the same issue with EMQX 5.7.1 running as podman container from the official OSS image.
Less retained messages: 

    emqx@f78374ae79d6:/opt/emqx$ emqx ctl retainer info
    Number of retained messages: 385

And the observer status:

<img width="1262" height="596" alt="Image" src="https://github.com/user-attachments/assets/72f97065-5656-46bb-b131-3d508c931483" />

<img width="2309" height="124" alt="Image" src="https://github.com/user-attachments/assets/a48be844-1d72-47a2-951e-07ac5349da43" />

Memory consumption in the container looks okaish:

<img width="851" height="357" alt="Image" src="https://github.com/user-attachments/assets/9b29ce0f-affb-489a-bbbf-e1cfbb62a7e6" />

**Comment by ingoratsdorf** [comment]
It's been a long time since I commented here last.
As per my last comment, I did stop the use of EMQX and moved to Mosquitto. But shortly thereafter I also moved all to proxmox and now run EMQX in an Ubuntu LXC. Rock solid, no spikes and weird memory issues.
Uses 104MB memory constant. Been like that for more than a year.

**So either a lot of improvement happened in EMQX or it's the docker environment.**

<img width="1230" height="335" alt="Image" src="https://github.com/user-attachments/assets/1f01aa23-6ee9-47ce-a757-063e249f2d57" />
